File: src/Quaero.Core/Storage/DataSourceStore.cs
using System.Text.Json;
using Quaero.Core.Models;

namespace Quaero.Core.Storage;

/// <summary>
/// Persists data source configurations to a JSON file.
/// This is the indexer configuration file shared between the UI and the Indexer service.
/// Runtime status (last run, doc count) is tracked in IndexRunLog in SQLite, not here.
/// </summary>
public class DataSourceStore
{
    private readonly string _filePath;
    private List<DataSource> _dataSources = new();
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public DataSourceStore(string? filePath = null)
    {
        _filePath = filePath ?? Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "Quaero", "datasources.json");

        var dir = Path.GetDirectoryName(_filePath);
        if (!string.IsNullOrEmpty(dir))
            Directory.CreateDirectory(dir);

        Load();
    }

    public string FilePath => _filePath;
    public IReadOnlyList<DataSource> DataSources => _dataSources.AsReadOnly();

    public void Add(DataSource dataSource)
    {
        _dataSources.Add(dataSource);
        Save();
    }

    public void Update(DataSource dataSource)
    {
        var index = _dataSources.FindIndex(ds => ds.Id == dataSource.Id);
        if (index >= 0)
        {
            _dataSources[index] = dataSource;
            Save();
        }
    }

    public void Remove(string id)
    {
        _dataSources.RemoveAll(ds => ds.Id == id);
        Save();
    }

    public DataSource? GetById(string id)
        => _dataSources.FirstOrDefault(ds => ds.Id == id);

    public List<DataSource> GetEnabled()
        => _dataSources.Where(ds => ds.Enabled).ToList();

    public void Reload() => Load();

    private void Load()
    {
        if (File.Exists(_filePath))
        {
            try
            {
                var json = File.ReadAllText(_filePath);
                _dataSources = JsonSerializer.Deserialize<List<DataSource>>(json, JsonOptions) ?? new();
            }
            catch
            {
                _dataSources = new();
            }
        }
    }

    private void Save()
    {
        var json = JsonSerializer.Serialize(_dataSources, JsonOptions);
        File.WriteAllText(_filePath, json);
    }
}

File: packages/core/src/data-source-store.ts
import fs from 'node:fs';
import path from 'node:path';
import { randomUUID } from 'node:crypto';
import { defaultDataSourcesPath } from './paths.js';
import type { DataSource } from './models.js';

/**
 * Persists data source configurations to a JSON file.
 * Runtime status is tracked in the SQLite run log, not here.
 */
export class DataSourceStore {
  private readonly filePath: string;
  private dataSources: DataSource[] = [];

  constructor(filePath?: string) {
    this.filePath = filePath ?? defaultDataSourcesPath();
    fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
    this.load();
  }

  get path(): string {
    return this.filePath;
  }

  list(): readonly DataSource[] {
    return this.dataSources;
  }

  getEnabled(): DataSource[] {
    return this.dataSources.filter((d) => d.enabled);
  }

  getById(id: string): DataSource | undefined {
    return this.dataSources.find((d) => d.id === id);
  }

  add(ds: Omit<DataSource, 'id'> & { id?: string }): DataSource {
    const created: DataSource = { id: ds.id ?? randomUUID().replace(/-/g, ''), ...ds };
    this.dataSources.push(created);
    this.save();
    return created;
  }

  update(ds: DataSource): boolean {
    const i = this.dataSources.findIndex((d) => d.id === ds.id);
    if (i < 0) return false;
    this.dataSources[i] = ds;
    this.save();
    return true;
  }

  remove(id: string): boolean {
    const before = this.dataSources.length;
    this.dataSources = this.dataSources.filter((d) => d.id !== id);
    const removed = this.dataSources.length !== before;
    if (removed) this.save();
    return removed;
  }

  reload(): void {
    this.load();
  }

  private load(): void {
    if (!fs.existsSync(this.filePath)) {
      this.dataSources = [];
      return;
    }
    try {
      const raw = fs.readFileSync(this.filePath, 'utf8');
      const parsed = JSON.parse(raw) as DataSource[];
      this.dataSources = Array.isArray(parsed) ? parsed : [];
    } catch {
      this.dataSources = [];
    }
  }

  private save(): void {
    fs.writeFileSync(this.filePath, JSON.stringify(this.dataSources, null, 2), 'utf8');
  }
}
